### ISSUE TITLE
Add documentation for Group Items in Processing Template

### ISSUE THREAD

**Issue by thomas-maschler** [Esri/raster-functions#29]
Please add documentation for Group Items in Processing Template to wiki.

When looking at RTF and python scripts alone (without the actual data), it is not clear how I can reference different raster datasets inside the raster mosaic as variables. 

How do I need to prepare the data/ raster mosaic table? How do I reference the different rasters in the template afterwards?

**Comment by jdrisdelle** [comment]
Hey Thomas, 

We are in the process of writing this document at the moment. Let me try to explain it here in the meantime. 

When applying  a function to  a stack of data in a mosaic dataset aggregation is controlled at the raster function template level. The properties of an rft can be accessed by right clicking on the top of the function chain and selecting **Properties** (in version 10.x).   

![image](https://cloud.githubusercontent.com/assets/8249074/13406962/fc64fcb0-deda-11e5-84e5-d7541209fd13.png)

The two settings in the rft properties that are used to aggregate records are the **tag field name** and **group field name**. you are also going to want to set the **type** parameter to **item group**   The two fields that you reference in these parameters also need to exist in your mosaic dataset attribute table. The records in the group field will contain a unique identifier that will represent a group of items in the mosaic dataset while the tag field name will reference items in those individual groups. For example: 

![image](https://cloud.githubusercontent.com/assets/8249074/13410415/816c3d5e-deec-11e5-9df7-4272d7c348d4.png)

When referencing these items in the Python raster function you are going to set the data type of that parameter to **rasters **.  You can see how this works in the [aggregate function](https://github.com/Esri/raster-functions/blob/master/functions/Aggregate.py). Items in the mosaic dataset will be passed into the python raster function as a group. You can then use this group of items do perform operations like [aggregation](https://github.com/Esri/raster-functions/blob/master/functions/Aggregate.py).  I hope this helps.  We will try to get the help topic up for this as soon as possible. Let me know if you have any questions.

**Comment by ThomasEmge** [comment]
Jamie,

I think you got the wrong Thomas ;)

**Comment by jdrisdelle** [comment]
Sorry @ThomasEmge . Not sure how that happened.

**Comment by thomas-maschler** [comment]
Thanks Jamie,

but how can I reference datasets in a specific group or with a specific tag inside the processing template?

By now I already figured out that if I write the tag in the Aliases field in the Variable Manager my raster function will only use rasters with that tag.

![image](https://cloud.githubusercontent.com/assets/7538337/13413599/b5bdaca0-df19-11e5-9f4f-4b0b38ca978f.png)

In the [windchill raster function template](https://github.com/Esri/raster-functions/blob/master/templates/Windchill.rft.xml) I saw you were using something like this 

`t@SFC`

What does that me? Can I reference a group like this?

Thanks,

Thomas

**Comment by akferoz** [comment]
@thomas-maschler, I'm sorry you had to figure that out without any/sufficient documentation. 

TL;DR: The string t@SFC is simply a tag associated with each temperature "slice" constructed by the GRIB raster type when used against an [NDFD](http://www.nws.noaa.gov/ndfd/) forecast [data](http://weather.noaa.gov/pub/SL.us008001/ST.opnl/DF.gr2/DC.ndfd/AR.conus/VP.004-007/) from the NWS. 
 
Here's one way to think about it…

[Windchill](https://github.com/Esri/raster-functions/blob/master/functions/Windchill.py) is a good example of a binary raster function--needing two input rasters, temperature and windspeed--housed in a [template](https://github.com/Esri/raster-functions/blob/master/templates/Windchill.rft.xml) with two variables of type raster as input. 

Unary functions like [NDVI](https://github.com/Esri/raster-functions/blob/master/functions/NDVI.py) or [KeyMetadata](https://github.com/Esri/raster-functions/blob/master/functions/KeyMetadata.py) are even simpler needing only one raster along with one or more scalars as input. 

A third category of function (based on cardinality) is one that expects an (unordered) stack of rasters, like the [Aggregate](https://github.com/Esri/raster-functions/blob/master/functions/Aggregate.py) function. 

In addition to the notes from @jdrisdelle, these are the points I want to emphasize when we document the interaction between functions, templates, map, mosaic dataset, and the user. 

- On one end, raster functions only define what their input expectations are in terms of data type and default value, using the `getParameterInfo` method. Raster functions do not (and cannot) determine how datasets or layers are eventually "connected" to those inputs. 
- On the other end, we have higher-level constructs that are containers of datasets like a map (containing image layers) or a mosaic dataset (containing items). As users, we are looking for ways to intuitively connect our raster function(s) with raster data. 
- Raster function _templates_ help make that connection. Exactly how ArcGIS uses your template is specific to the application and the context. Input raster datasets "bind" to input parameters of raster functions through _variables_ defined in the template. More precisely, a raster dataset identified as, say, R, will bind with a particular template variable if it's named R or if it has an alias R. 

In ArcMap, raster-based layers (like image server layer, mosaic layer and raster layer) are named based on their order in the Table of Contents. So, the first layer is `Raster1`, the second one is `Raster2`, and so on. This enables us to use any template (driven by any binary function) as long as the variable name (or alias) reads `Raster1` and `Raster2`. Additionally, the first layer also goes by the name `Raster` to enable unary templates to work correctly. To enable aggregation, the virtual stack of all selected rasters is called `Raster[]`. We use order-based naming here because other methods of tagging the layers (based on, say, the layer name) would severely diminish your function templates' re-usability.  

Mosaic datasets add another level of complexity because we now have multiple groups of rasters against which the template needs to be repeatedly applied in order to construct the mosaicked image. We can use a field in the mosaic dataset to uniquely identify different groups of items—think of them as neighborhoods or subsets in your collection. As @jdrisdelle described, this field is indicated to the function template's Group Field Name parameter. We use another field—specified in the Tag Field Name parameter—to identify each item within a particular group. The item's tag is its name and is used for binding with template variables (as long as it matches the variable's name or one of the aliases). We can use this construct to group together multiple items (like temperature and windspeed) by, say, time. Analogous to the map-based scenario, all items in a group are made available as a virtual stack and tagged `Raster[]`. This enables aggregation-type templates to work as well. Because a mosaic dataset is carefully authored and its items managed as records in a database table, we avoid order-based tagging of items and resort to using author-define or raster-type-specified names. Lastly, these rules for a mosaic dataset are in addition to prevailing constraints on each item w.r.t. visibility- and query-based filtering. 


Hope that makes sense.
